sdks/go/pkg/beam/registration/emitterIterRegistration.go

```diff
@@ -0,0 +1,298 @@
+			gen.fn = gen.invoke
+			return gen
+		}
+	}
+	exec.RegisterEmitter(reflect.TypeOf(e).Elem(), registerFunc)
+}
+
+// RegisterEmitter3 registers parameters from your DoFn with a
+// signature func(T1, T2, T3) and optimizes their execution.
+// This must be done by passing in a reference to an instantiated version
+// of the function, aka:
+// registration.RegisterEmitter3[T1, T2, T3]((*func(T1, T2, T3))(nil))
+func RegisterEmitter3[T1, T2, T3 any](e *func(T1, T2, T3)) {
+	registerFunc := func(n exec.ElementProcessor) exec.ReusableEmitter {
+		gen := &emitNative2WithTimestamp[T2, T3]{n: n}
+		gen.fn = gen.invoke
+		return gen
+	}
+	exec.RegisterEmitter(reflect.TypeOf(e).Elem(), registerFunc)
+}
+
+type iterNative1[T any] struct {
+	s  exec.ReStream
+	fn interface{}
+
+	// cur is the "current" stream, if any.
+	cur exec.Stream
+}
+
+func (v *iterNative1[T]) Init() error {
+	cur, err := v.s.Open()
+	if err != nil {
+		return err
+	}
+	v.cur = cur
+	return nil
+}
+
+func (v *iterNative1[T]) Value() interface{} {
+	return v.fn
+}
+
+func (v *iterNative1[T]) Reset() error {
+	if err := v.cur.Close(); err != nil {
+		return err
+	}
+	v.cur = nil
+	return nil
+}
+
+func (v *iterNative1[T]) invoke(value *T) bool {
+	elm, err := v.cur.Read()
+	if err != nil {
+		if err == io.EOF {
+			return false
+		}
+		panic(fmt.Sprintf("broken stream: %v", err))
+	}
+	*value = elm.Elm.(T)
+	return true
+}
+
+type iterNative2[T1, T2 any] struct {
+	s  exec.ReStream
+	fn interface{}
+
+	// cur is the "current" stream, if any.
+	cur exec.Stream
+}
+
+func (v *iterNative2[T1, T2]) Init() error {
+	cur, err := v.s.Open()
+	if err != nil {
+		return err
+	}
+	v.cur = cur
+	return nil
+}
+
+func (v *iterNative2[T1, T2]) Value() interface{} {
+	return v.fn
+}
+
+func (v *iterNative2[T1, T2]) Reset() error {
+	if err := v.cur.Close(); err != nil {
+		return err
+	}
+	v.cur = nil
+	return nil
+}
+
+func (v *iterNative2[T1, T2]) invoke(key *T1, value *T2) bool {
+	elm, err := v.cur.Read()
+	if err != nil {
+		if err == io.EOF {
+			return false
+		}
+		panic(fmt.Sprintf("broken stream: %v", err))
+	}
+	*key = elm.Elm.(T1)
+	*value = elm.Elm2.(T2)
+	return true
+}
+
+// RegisterIter1 registers parameters from your DoFn with a
+// signature func(*T) bool and optimizes their execution.
+// This must be done by passing in a reference to an instantiated version
+// of the function, aka:
+// registration.RegisterIter1[T]((*func(*T) bool)(nil))
+func RegisterIter1[T any](i *func(*T) bool) {
+	registerFunc := func(s exec.ReStream) exec.ReusableInput {
+		ret := &iterNative1[T]{s: s}
+		ret.fn = ret.invoke
+		return ret
+	}
+	exec.RegisterInput(reflect.TypeOf(i).Elem(), registerFunc)
+}
+
+// RegisterIter1 registers parameters from your DoFn with a
+// signature func(*T1, *T2) bool and optimizes their execution.
+// This must be done by passing in a reference to an instantiated version
+// of the function, aka:
+// registration.RegisterIter2[T1, T2]((*func(*T1, *T2) bool)(nil))
```

Comment:
What do you think about dropping the function type parameter?

1. Users aren't going to have an iterator/emitter around, so they'll need to specify func pointer set up like you've got. 
2. Even assuming type inference worked, it's still annoying to go through the awkward (*blah)(nil) dance.

https://go.dev/play/p/quG3c5scpGQ

Inference does seem to work at least. We do get compile time safety with this approach too, but it also prevents us from automatically supporting any "new" iterator or emitter formats in the future, with the same registration calls.

What do you think?

Reply:
I think that's a great idea. Updated to do this!

> it also prevents us from automatically supporting any "new" iterator or emitter formats in the future, with the same registration calls.

We don't really get that right now since we're depending on parameters being in the correct order/placement with the current approach.